app/vue/src/client/preview/util.ts

```diff
@@ -1,23 +1,38 @@
 import { VueConstructor } from 'vue';
+import { PropValidator, PropOptions } from 'vue/types/options';
 
-function getType(fn: Function) {
-  const match = fn && fn.toString().match(/^\s*function (\w+)/);
+function getType(fn?: Function): string {
+  const match = fn?.toString().match(/^\s*function (\w+)/);
   return match ? match[1] : '';
 }
 
-// https://github.com/vuejs/vue/blob/dev/src/core/util/props.js#L92
-function resolveDefault({ type, default: def }: any) {
-  if (typeof def === 'function' && getType(type) !== 'Function') {
+/**
+ * Extracts the default value of a prop
+ * Also see https://github.com/vuejs/vue/blob/dev/src/core/util/props.js#L92
+ */
+function resolveDefault<T>(validator: PropValidator<T>): T | null | undefined {
+  if (typeof validator === 'function' || Array.isArray(validator)) {
+    return undefined;
+  }
+
+  const canTakeFunctionProp = Array.isArray(validator.type)
+    ? validator.type.some((t) => getType(t) === 'Function')
+    : getType(validator.type) === 'Function';
+
+  if (typeof validator.default === 'function' && canTakeFunctionProp) {
     // known limitation: we don't have the component instance to pass
-    return def.call();
+    return (validator.default as () => T | null | undefined)();
```

Comment:
Tried as I might, I couldn't get this to be completely type-safe using a type guard. I think this would need [negated types](https://github.com/microsoft/TypeScript/pull/29317) to remove `T & Function` from the possible types.